src/Data/Pool/Internal.hs

```diff
@@ -71,8 +76,20 @@ data PoolConfig a = PoolConfig
   -- capabilities and rounded up. Therefore the pool might end up creating up to
   -- @N - 1@ resources more in total than specified, where @N@ is the number of
   -- capabilities.
+  , poolTimeoutConfig :: Maybe TimeoutConfig
+  -- ^ Optional timeout for waiting for a resource
+  }
+
+data TimeoutConfig = TimeoutConfig
+  { acquireResourceTimeout :: Int
+  -- ^ Time to await, microseconds
+  , timeoutLabel :: String 
```

Comment:
Wouldn't it be better to provide a value that should be thrown as exception here instead of a label?

Then you can also get rid of the `TimeoutException`.